Data/HashMap/Base.hs

```diff
@@ -751,12 +1083,131 @@ update f = alter (>>= f)
 -- absence thereof. @alter@ can be used to insert, delete, or update a value in a
 -- map. In short : @'lookup' k ('alter' f k m) = f ('lookup' k m)@.
 alter :: (Eq k, Hashable k) => (Maybe v -> Maybe v) -> k -> HashMap k v -> HashMap k v
+-- TODO(m-renaud): Consider using specialized insert and delete for alter.
 alter f k m =
   case f (lookup k m) of
     Nothing -> delete k m
     Just v  -> insert k v m
 {-# INLINABLE alter #-}
 
+-- | /O(log n)/  The expression (@'alterF' f k map@) alters the value @x@ at
+-- @k@, or absence thereof. @alterF@ can be used to insert, delete, or update
+-- a value in a map.
+--
+-- Note: 'alterF' is a flipped version of the 'at' combinator from
+-- <https://hackage.haskell.org/package/lens-4.15.4/docs/Control-Lens-At.html#v:at Control.Lens.At>.
+--
+-- @since 0.2.9
+alterF :: (Functor f, Eq k, Hashable k)
+       => (Maybe v -> f (Maybe v)) -> k -> HashMap k v -> f (HashMap k v)
+-- We only calculate the hash once, but unless this is rewritten
+-- by rules we may test for key equality multiple times.
+-- We force the value of the map for consistency with the rewritten
+-- version; otherwise someone could tell the difference using a lazy
+-- @f@ and a functor that is similar to Const but not actually Const.
+alterF f !k !m = (<$> f mv) $ \fres ->
+  case fres of
+    Nothing -> delete' h k m
+    Just v' -> insert' h k v' m
+
+  where !h = hash k
+        mv = lookup' h k m
+{-# INLINABLE [0] alterF #-}
+
+#if MIN_VERSION_base(4,8,0)
+bogus :: a
+bogus = error "Data.HashMap.alterF internal error: hit bottom"
+
+bogus# :: (# #) -> (# a #)
+bogus# _ = error "Data.HashMap.alterF internal error: hit bottom#"
+
+{-# RULES
+"alterFWeird" forall f k m. alterF f k m = alterFWeird (f Nothing) (f (Just bogus)) f k m
```

Comment:
Oh here it is, I missed this when reading it the first time. Can you add a comment that we unconditionally rewrite alterF? Also consider dropping the rewrite part `alterF f k m = ...` to the next line so its separated visually from the forall part.